File: webapp/models/location.ts
export interface IBase {
    uid?: string;
    name?: string;
    code?: string;
}

export interface ICountry extends IBase {
}

export interface ILocation extends IBase {
    email?: string;
    emailCc?: string[];
    consentEmail?: boolean;
    businessPhone?: string;
    mobilePhone?: string;
    consentSms?: boolean;
}

export interface IProvince extends ILocation {
    country?: ICountry;
    countryUid?: string;
}

export interface IDistrict extends ILocation {
    province?: IProvince;
    provinceUid?: string;
}

export interface IGenericLocation extends ILocation {
    district?: IDistrict;
    districtUid?: string;
    province?: IProvince;
    provinceUid?: string;
    country?: ICountry;
    countryUid?: string;
}

File: felicity/apps/setup/entities/location.py
from sqlalchemy import Boolean, Column, ForeignKey, String
from sqlalchemy.orm import relationship

from felicity.apps.abstract import BaseEntity


class LocationBase(BaseEntity):
    __abstract__ = True
    code = Column(String, index=True, unique=True, nullable=True)
    name = Column(String)
    email = Column(String, nullable=True)
    email_cc = Column(String, nullable=True)
    mobile_phone = Column(String, nullable=True)
    business_phone = Column(String, nullable=True)
    active = Column(Boolean(), default=False)


class District(LocationBase):
    __tablename__ = "district"

    province_uid = Column(String, ForeignKey("province.uid"))
    province = relationship("Province", backref="districts", lazy="selectin")


class Province(LocationBase):
    __tablename__ = "province"

    country_uid = Column(String, ForeignKey("country.uid"))
    country = relationship("Country", backref="provinces", lazy="selectin")


class Country(BaseEntity):
    __tablename__ = "country"

    name = Column(String, default="Zimbabwe")
    code = Column(String, index=True, unique=True, nullable=True)
    active = Column(Boolean(), default=False)
